libs/unity/library/Runtime/Scripts/Media/MediaLine.cs

```diff
@@ -141,36 +156,48 @@ public MediaTrack SenderTrack
         /// If <see cref="Transceiver"/> is valid, that is a first session negotiation has already been conducted,
         /// then changing this value raises a <see xref="WebRTC.PeerConnection.RenegotiationNeeded"/> event on the
         /// peer connection of <see cref="Transceiver"/>.
-        /// </summary>
+        ///
+        /// Must be changed on the main thread.
+        /// </remarks>
         public MediaReceiver Receiver
         {
             get { return _receiver; }
             set
             {
-                if (value == null)
+                if (_receiver == value)
                 {
-                    if (_receiver != null)
-                    {
-                        _receiver.OnRemovedFromMediaLine(this);
-                        _receiver = null;
-                    }
+                    return;
                 }
-                else
+                if (value!= null && value.MediaKind != MediaKind)
                 {
-                    if (value.MediaKind != MediaKind)
+                    throw new ArgumentException("Wrong media kind", nameof(Receiver));
+                }
+
+                // If the connection is active, ensure that this is run on the main thread
+                // and doesn't race with the signaling handlers.
+                // If the connection is not awake we cannot use EnsureIsMainAppThread, but
+                // there should be no race on shared state, so skip the check.
+                if (_peer.isActiveAndEnabled)
+                {
+                    _peer.EnsureIsMainAppThread();
+                }
```

Comment:
```suggestion
                _peer.EnsureIsMainAppThread();
```

Reply:
The issue here is that `EnsureIsMainAppThread` will throw if called before the PC is awake (which will happen normally - it happens in tests now - if the PC is added to an inactive object), but setting Source/Receiver on a "sleeping" PC should be fine.

If `isActiveAndEnabled` throws consistently we can use that in `EnsureIsMainAppThread` rather than checking the (possibly uninitialized) local thread variable.

Reply:
I don't know if it throws consistently, I would rely on consistently doing something forbidden by Unity and expecting Unity will catch that and throw.

In any case the current code doesn't work, because either `isActiveAndEnabled` is accessed from a background thread and it's garbage/inconsistent/throws, so the test `EnsureIsMainAppThread()` might not even run, or it is from the main Unity app thread, and the test `EnsureIsMainAppThread()` is useless.

If there's no better way, then just delete everything and don't test. Nobody expects users to be able to modify Unity component properties from a background thread anyway, so it's mostly about making sure our code does the right thing.

Reply:
The purpose of that code was to throw if called from outside the app thread and not throw when called on the app thread, which it served perfectly. `isActiveAndEnabled` does throw, it seems. I could have just called that and discarded the result, but I felt the above seemed more readable, even though I agree it looks weird at an accurate analysis.

Anyway I agree we shouldn't rely on a specific behavior of `isActiveAndEnabled`, so I have rearranged/removed the checks.